drivers: sensors: Implement MAX31865 sensor

maass-hamburg opened the pull request:
This commit implements the temperature sensor interface for
the Maxim MAX31865 SPI Temperature Sensor. It also provides a sample.

Signed-off-by: Fin Maaß <fin.maass@haw-hamburg.de>

galak commented on dts/bindings/sensor/maxim,max31865.yaml:
```diff
@@ -0,0 +1,47 @@
+        required: false
+        description: High fault threshold (ADC CODE)
+
+  three-wire:
```
use `maxim,3-wire` as property name to match "upstream" linux binding:

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/iio/temperature/maxim,max31865.yaml

galak reviewed (changes requested):
* a fair amount of extra whitespace in driver needs cleaning up
* please add to `tests/drivers/build_all/sensor`
* minor fix for dts binding
* Do we need a unique sample for this sensor?
* only glanced over the driver code, is newlib truly needed for this driver?

maass-hamburg commented:
> * Do we need a unique sample for this sensor?
> * only glanced over the driver code, is newlib truly needed for this driver?

Most sensor drivers have their own sample so this drive should also have its own.
The newlib is necessary for the `sqrt` funktion in the max31865.c to calculate the temperature.

MaureenHelm commented on drivers/sensor/max31865/Kconfig:
```diff
@@ -0,0 +1,10 @@
+
+config MAX31865
+	bool "MAX31865 Temperature Sensor"
+	depends on SPI && NEWLIB_LIBC
```
```suggestion
	depends on NEWLIB_LIBC && DT_HAS_MAXIM_MAX31865_ENABLED
	select SPI
```

MaureenHelm commented on drivers/sensor/max31865/max31865.c:
```diff
@@ -0,0 +1,299 @@
+ */
+#include "max31865.h"
+
+int8_t initMax31865(const struct device *dev)
```
Please no camel case

MaureenHelm commented on drivers/sensor/max31865/max31865.c:
```diff
@@ -0,0 +1,299 @@
+	read_reg.u16 = 0;
+	double resistance = 0;
+
+	k_sleep(K_MSEC(128));
```
Please document why this value was chosen

MaureenHelm commented on dts/bindings/sensor/maxim,max31865.yaml:
```diff
@@ -0,0 +1,47 @@
+        type: int
+        default: 32767
+        required: false
+        description: High fault threshold (ADC CODE)
```
Property defaults need justification in the description.

MaureenHelm reviewed (changes requested):
Please squash the requested changes into the original commit, and rebase to latest main to remove the merge commit.

MaureenHelm commented:
> Please squash the requested changes into the original commit, and rebase to latest main to remove the merge commit.

This still needs to be done.

maass-hamburg commented:
> > Please squash the requested changes into the original commit, and rebase to latest main to remove the merge commit.
> 
> This still needs to be done.

It should be done now. 
With this last commit I also removed the dependency on newlib by implementing a less accurate  function which is used, when newlib is not enabled.

maass-hamburg commented:
> > Please squash the requested changes into the original commit, and rebase to latest main to remove the merge commit.
> 
> This still needs to be done.

It should be done now. This is my first commit to Zephyr and I am not that used to work with some git functions.

MaureenHelm commented on drivers/sensor/max31865/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+# Makefile - MAX31865
```
This isn't a Makefile

MaureenHelm commented on drivers/sensor/max31865/max31865.c:
```diff
@@ -0,0 +1,318 @@
+ */
+#include "max31865.h"
+
+int initmax31865(const struct device *dev)
```
All of the functions in this file should be static
```suggestion
static int init_max31865(const struct device *dev)
```

MaureenHelm commented on drivers/sensor/max31865/max31865.c:
```diff
@@ -0,0 +1,318 @@
+	const struct max31865_config *cfg = dev->config;
+
+	if (!spi_is_ready(&cfg->spi)) {
+		return -1;
```
```suggestion
		return -ENODEV;
```

MaureenHelm commented on drivers/sensor/max31865/max31865.c:
```diff
@@ -0,0 +1,318 @@
+
+	ret = spi_transceive_dt(&cfg->spi, &tx, &rx);
+	if (ret) {
+		LOG_DBG("spi_transceive FAIL %d\n", ret);
```
```suggestion
		LOG_ERR("spi_transceive FAIL %d\n", ret);
```

MaureenHelm commented on drivers/sensor/max31865/max31865.h:
```diff
@@ -0,0 +1,106 @@
+	uint16_t high_threshold;        /**< High fault threshold. */
+};
+
+/* Bit manipulation macros */
+#define SETBIT(x, y)        (x) |= BIT(y)
+#define CLEARBIT(x, y)      (x) &= ~BIT(y)
+#define TESTBIT(data, pos)  ((0u == (data & BIT(pos)))?0u:1u)
+#define SWAP(a, b) (((a) ^= (b)), ((b) ^= (a)), ((a) ^= (b)))
```
Please use macros from `include/zephyr/sys/util*.h` instead

MaureenHelm commented on drivers/sensor/max31865/max31865.c:
```diff
@@ -0,0 +1,318 @@
+		/*Clear fault register */
+		data->config_control_bits |= 0x02;
+		configure_device(dev);
+		LOG_WRN("Fehler: %10x", fault_register);
```
English please

MaureenHelm commented on drivers/sensor/max31865/max31865.c:
```diff
@@ -0,0 +1,318 @@
+	set_threshold_values(dev);
+}
+
+int max31865_getTemperatureSensor(const struct device *dev, struct sensor_value *val)
```
```suggestion
static int max31865_get_temperature_sensor(const struct device *dev, struct sensor_value *val)
```

MaureenHelm commented on drivers/sensor/max31865/max31865.h:
```diff
@@ -0,0 +1,106 @@
+#define TESTBIT(data, pos)  ((0u == (data & BIT(pos)))?0u:1u)
+#define SWAP(a, b) (((a) ^= (b)), ((b) ^= (a)), ((a) ^= (b)))
+
+void max31865_init(const struct device *dev);
+void max31865_set_vbias(const struct device *dev, bool enable);
+int max31865_get_temperature(const struct device *dev, double *temperature);
+int max31865_spi_write(const struct device *dev, uint8_t reg, uint8_t *data, uint8_t lenght);
+int initmax31865(const struct device *dev);
+int max31865_getTemperatureSensor(const struct device *dev, struct sensor_value *val);
+void configure_device(const struct device *dev);
+void set_threshold_values(const struct device *dev);
+double calculate_temperature(double resistance, double resistance_0);
```
Remove

MaureenHelm commented on dts/bindings/sensor/maxim,max31865.yaml:
```diff
@@ -0,0 +1,47 @@
+
+properties:
+
+  resistance-at-zero:
+        type: int
+        default: 100
+        required: false
+        description: Sensor resistance at 0 Celsius (100 for PT100, 1000 for PT1000)
+
+  resistance-reference:
+        type: int
+        default: 430
+        required: false
+        description: Circuit reference resistance (recommended on MAX31865 datasheet 400 ohms for PT100, 4000 for PT1000, the Adafruit boards use 430 ohms and 4300 ohms)
```
These should be required, since we can't assume what is on the board.

MaureenHelm commented on samples/sensor/max31865/README:
```diff
@@ -0,0 +1,3 @@
+This is a sample app to read a MAX31865.
```
This file should be in .rst format. See `samples/hello_world/README.rst` for example

MaureenHelm commented on samples/sensor/max31865/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
```
Please move the sample changes to a separate commit in this PR. Please also add a `sample.yaml`

maass-hamburg committed:
drivers: sensors: Implement MAX31865 sensor

This commit implements the temperature sensor interface for
the Maxim MAX31865 SPI Temperature Sensor.

Signed-off-by: Fin Maaß <fin.maass@haw-hamburg.de>

2

dcpleung commented on samples/sensor/max31865/prj.conf:
```diff
@@ -0,0 +1,11 @@
+CONFIG_PRINTK=y
+CONFIG_CONSOLE=y
+CONFIG_USE_SEGGER_RTT=y
+CONFIG_RTT_CONSOLE=y
+CONFIG_UART_CONSOLE=n
+CONFIG_SHELL=y
+CONFIG_SHELL_BACKEND_RTT=y
+CONFIG_SERIAL=n
```
Does this require shell to work properly? If not, please remove `CONFIG_SHELL`.

Also, This looks like a generic sample where any board with a SPI bus can be used. So I think it is better to move the config to enabling RTT and disabling UART to a board overlay (under this sample). This will allow other boards with different serial/console to use this sample.

dcpleung commented on samples/sensor/max31865/sample.yaml:
```diff
@@ -0,0 +1,10 @@
+sample:
+  name: MAX31865 temperature sensor
+tests:
+  sample.sensor.max31865:
+    harness: sensor
+    tags: samples sensors
+    depends_on: spi
+    platform_allow: particle_xenon
```
Remove the `platform_allow` so this is not restricted to one board.

And add `filter: dt_compat_enabled("maxim,max31865")` so this can be built (via twister) any boards.

dcpleung commented on drivers/sensor/max31865/max31865.c:
```diff
@@ -0,0 +1,302 @@
+	}
+}
+
+#ifdef NEWLIB_LIBC
```
This needs to be `CONFIG_NEWLIB_LIBC`.

dcpleung reviewed (changes requested):
BTW, please reverse the order of the commit so the introduction of the driver comes first.

gmarull commented on samples/sensor/max31865/src/main.c:
```diff
@@ -0,0 +1,38 @@
+		if (r != 0) {
+			continue;
+		}
+		sensor_channel_get(max_sensor, SENSOR_CHAN_AMBIENT_TEMP, &sv.str);
```
let's not add yet another sensor sample without being generic for the sensor type (unless it's really needed to demonstrate special features). See the work done for the generic accelerometer polling sample.

maass-hamburg replied:
So should the sample be removed?

maass-hamburg commented:
I removed the sample. It is still available in another branch [https://github.com/FTZ-DiWiP/zephyr/tree/max31865_sample](url) and could be added later, if wished. But this pull request now only provides the driver without the sample.

MaureenHelm commented on drivers/sensor/max31865/max31865.h:
```diff
@@ -0,0 +1,98 @@
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/gpio.h>
```
Is this needed? I don't see any GPIO operations in the driver

MaureenHelm commented on drivers/sensor/max31865/max31865.h:
```diff
@@ -0,0 +1,98 @@
+#include <zephyr/kernel.h>
+LOG_MODULE_REGISTER(MAX31865, CONFIG_SENSOR_LOG_LEVEL);
+
+#define MAX31865_FAULT_HIGH_THRESHOLD (1 << 7)
+#define MAX31865_FAULT_LOW_THRESHOLD  (1 << 6)
+#define MAX31865_FAULT_REFIN	      (1 << 5)
+#define MAX31865_FAULT_REFIN_FORCE    (1 << 4)
+#define MAX31865_FAULT_RTDIN_FORCE    (1 << 3)
+#define MAX31865_FAULT_VOLTAGE	      (1 << 2)
```
Please use `BIT()`

MaureenHelm commented on drivers/sensor/max31865/max31865.h:
```diff
@@ -0,0 +1,98 @@
+
+struct max31865_data {
+	struct sensor_value val;
+	uint8_t config_control_bits;
+	uint16_t resistance_at_zero;
+	uint16_t resistance_reference;
```
None of these appear to ever change at runtime, so there's no reason to keep copies in this struct. Just access them from the config struct directly

maass-hamburg replied:
`config_control_bits` get changed at runtime, but the others don't.

MaureenHelm commented on dts/bindings/sensor/maxim,max31865.yaml:
```diff
@@ -0,0 +1,44 @@
+  resistance-reference:
+        type: int
+        required: true
+        description: Circuit reference resistance (recommended on MAX31865 datasheet 400 ohms for PT100, 4000 for PT1000, the Adafruit boards use 430 ohms and 4300 ohms)
```
Units?

MaureenHelm commented on dts/bindings/sensor/maxim,max31865.yaml:
```diff
@@ -0,0 +1,44 @@
+        type: int
+        default: 0
+        required: false
+        description: Low fault threshold (ADC CODE, 15-bit value, default value is the minimum value)
```
Units?

maass-hamburg commented:
So any other requested changes?

MaureenHelm commented on drivers/sensor/max31865/max31865.c:
```diff
@@ -0,0 +1,304 @@
+		LOG_ERR("SPI read %d\n", err);
+		return -EIO;
+	}
+	byteswp(read_reg.u8, read_reg.u8 + 1, 1);
```
I think you want `sys_be16_to_cpu` instead

MaureenHelm commented on drivers/sensor/max31865/max31865.c:
```diff
@@ -0,0 +1,304 @@
+		.high_threshold = DT_INST_PROP(inst, high_threshold),                              \
+	};                                                                                         \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(inst, max31865_init, NULL, &max31865_data_##inst,                    \
```
Use `SENSOR_DEVICE_DT_INST_DEFINE` instead
